src/repo/kv.rs

```diff
@@ -0,0 +1,387 @@
+                let key = get_pin_key(target, &mode);
+                batch.remove(key.as_str());
+            }
+            _ => {}
+        }
+
+        let recursive_key = get_pin_key(target, &PinMode::Recursive);
+        batch.insert(recursive_key.as_str(), "");
+
+        for cid in &set {
+            let indirect_key = get_pin_key(cid, &PinMode::Indirect);
+
+            let is_already_pinned = is_pinned(self, cid);
+
+            if let Ok(true) = is_already_pinned {
+                continue;
+            }
+
+            // value is for get information like "Qmd9WDTA2Kph4MKiDDiaZdiB4HJQpKcxjnJQfQmM5rHhYK indirect through QmXr1XZBg1CQv17BPvSWRmM7916R6NLL7jt19rhCPdVhc5"
+            batch.insert(indirect_key.as_str(), target.to_string().as_str());
+        }
+
+        Ok(self._apply_batch(batch)?)
+    }
+
+    async fn remove_direct_pin(&self, target: &Cid) -> Result<(), Error> {
+        if is_not_pinned_or_pinned_indirectly(self, target)? {
+            return Err(anyhow::anyhow!("not pinned or pinned indirectly"));
+        }
+
+        let key = get_pin_key(target, &PinMode::Direct);
+
+        Ok(self._remove(&key)?)
+    }
+
+    async fn remove_recursive_pin(
+        &self,
+        target: &Cid,
+        referenced: References<'_>,
+    ) -> Result<(), Error> {
+        let set = referenced
+            .try_collect::<std::collections::BTreeSet<_>>()
+            .await?;
+
+        if is_not_pinned_or_pinned_indirectly(self, target)? {
+            return Err(anyhow::anyhow!("not pinned or pinned indirectly"));
+        }
+
+        let mut batch = sled::Batch::default();
+
+        let recursive_key = get_pin_key(target, &PinMode::Recursive);
+
+        batch.remove(recursive_key.as_str());
+
+        for cid in &set {
+            let already_pinned = get_pinned_mode(self, cid)?;
+
+            match already_pinned {
+                Some(PinMode::Recursive) | Some(PinMode::Direct) => continue, // this should be unreachable
+                Some(PinMode::Indirect) => {
+                    let indirect_key = get_pin_key(cid, &PinMode::Indirect);
+                    batch.remove(indirect_key.as_str());
+                }
+                _ => {}
+            }
+        }
+
+        Ok(self._apply_batch(batch)?)
+    }
+
+    async fn list(
+        &self,
+        expected_mode: Option<PinMode>,
+    ) -> futures::stream::BoxStream<'static, Result<(Cid, PinMode), Error>> {
+        let db = self.get_db();
+
+        // the minimum cid of version 0
+        let min_key = "pin.0.0000000000000000000000000000000000000000000000";
+        assert_eq!(min_key.len(), 52);
+
+        let iter = db.range(min_key..);
+        let mut all_keys: Vec<String> = vec![];
+
+        for item in iter {
+            if item.is_err() {
+                continue;
+            }
+
+            let (raw_key, _) = item.unwrap();
+            let key = String::from(String::from_utf8_lossy(raw_key.as_ref()));
```

Comment:
the outer conversion is redundant

Reply:
Good advice, the result of `String::from_utf8_lossy(raw_key.as_ref())` (is a Cow<'_, str>) should be converted String at the push stage, but can the result of `.into_owned` is a `String`?

Reply:
yes: `Cow::into_owned` will create a `String` in this case

Reply:
the filter below can be done before the lossy utf-8 conversion, but I was thinking that perhaps the whole of the str conversion could be perhaps avoided.

Reply:
yeah, I was wondering the same; strings are always UTF-8 validated in Rust, and it's not necessarily needed here